### PR TITLE
Update VM code & setting for RISC-V in OpenJ9 (part6/VM)

### PR DESCRIPTION
**Pull request by ChengJin01**
The changes mainly update & add interpreter specific code
(with placeholders to be replaced by JIT) & setting
to enable on RISC-V 64bit from the OpenJ9 perspective.

Issue: eclipse#7421

Signed-off-by: Cheng Jin <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
Already verified the compilation on Fedora_Linux/QEMU.

Note: The JIT related code (on register setting) util/jitregs.c was written against the RISC-V Spec (please refer to https://content.riscv.org/wp-content/uploads/2017/05/riscv-spec-v2.2.pdf for details)
and should be modified by JIT to ensure it works correctly from JIT perspective.

In addition, rv64helpers.m4 & rv64cinterp.m4 should be updated by JIT later to insert JIT frames so
as to replace rv64cinterpnojit.cpp(the wrapper of interpreter for now without JIT)

Reviewer: @gacholio  , @pshipton , @DanHeidinga

File: runtime/vm/module.xml [hunk 1]
```diff
@@ -90,6 +90,12 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 			<makefilestub data="UMA_OBJECTS:=$(filter-out zcinterp%,$(UMA_OBJECTS))\n">
 				<exclude-if condition="spec.flags.arch_s390"/>
 			</makefilestub>
+			<makefilestub data="UMA_OBJECTS:=$(filter-out rv64cinterp%,$(UMA_OBJECTS))\n">
+				<exclude-if condition="spec.flags.arch_riscv and spec.flags.J9VM_ENV_DATA64"/>
+			</makefilestub>
+			<makefilestub data="UMA_OBJECTS:=$(filter-out rv64cinterpnojit%,$(UMA_OBJECTS))\n">
```
Comment (keithc-ca) [inline runtime/vm/module.xml]: I don't think the `\n` is necessary nor desirable in the stubs.

File: runtime/oti/j9nonbuilder.h [hunk 2]
```diff
@@ -5135,7 +5201,7 @@ typedef struct J9JavaVM {
 	omrthread_monitor_t systemPropertiesMutex;
 	U_8* javaHome;
 	void* cInterpreter;
-	void* bytecodeLoop;
+	void (*bytecodeLoop)(struct J9VMThread *currentThread);
```
Comment (keithc-ca) [inline runtime/oti/j9nonbuilder.h]: According to `vm_api.h` [1] those functions have signature `UDATA (*)(J9VMThread *)` while the assignment in `jvminit.c` expects this field to be of type `void *` [2]: I don't expect this to compile without further changes.

[1] https://github.com/eclipse/openj9/blob/master/runtime/oti/vm_api.h#L4069-L4080
[2] https://github.com/eclipse/openj9/blob/master/runtime/vm/jvminit.c#L2414-L2416

File: runtime/oti/j9nonbuilder.h [hunk 3]
```diff
@@ -5698,6 +5764,11 @@ typedef struct J9CInterpreterStackFrame {
 	U_8 preservedFPRs[8 * 8]; /* v8-15 */
 	UDATA jitGPRs[32]; /* x0-x31 */
 	U_8 jitFPRs[32 * 8]; /* v0-v31 */
+#elif defined(J9VM_ARCH_RISCV) /* J9VM_ARCH_ARM */
+	UDATA preservedGPRs[13]; /* x2, x8, x9, and x18-x27  */
+	U_8 preservedFPRs[32]; /* f0-f31 */
```
Comment (keithc-ca) [inline runtime/oti/j9nonbuilder.h]: Should this be 8 times larger? `U_8 preservedFPRs[32 * 8];`

File: runtime/vm/module.xml (hunk 1 above)
Comment (ChengJin01) [reply runtime/vm/module.xml]: Fixed as suggested above.

File: runtime/oti/j9nonbuilder.h (hunk 2 above)
Comment (ChengJin01) [reply runtime/oti/j9nonbuilder.h]: I changed the existing code as discussed with @keithc-ca,  will compile a cross-build with the latest changes to ensure whether it works fine in the cross-compilation.

File: runtime/oti/j9nonbuilder.h (hunk 3 above)
Comment (ChengJin01) [reply runtime/oti/j9nonbuilder.h]: Updated. It might be modified accordingly in JIT based on their real requirements.

**Comment by keithc-ca** [comment]
@ChengJin01 Please resolve the merge conficts.

**Comment by ChengJin01** [comment]
@keithc-ca, the conflict has been fixed.

File: runtime/vm/rv64/unsafeHelper.s [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+# General Public License, version 2 with the GNU Classpath
+# Exception [1] and GNU General Public License, version 2 with the
+# OpenJDK Assembly Exception [2].
+#
+# [1] https://www.gnu.org/software/classpath/license.html
+# [2] http://openjdk.java.net/legal/assembly-exception.html
+#
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+
+	.text
```
Comment (keithc-ca) [inline runtime/vm/rv64/unsafeHelper.s]: Why is this indented with a tab while the rest of the file is indented with spaces?
Other linux architectures use tabs - I suggest this file and `rv64cinterp.m4` should use tabs as well.

Comment (ChengJin01) [reply runtime/vm/rv64/unsafeHelper.s:21]: Replaced these spaces with tabs as suggested.

File: runtime/vm/jvminit.c [hunk 5]
```diff
@@ -2432,8 +2432,8 @@ IDATA VMInitStages(J9JavaVM *vm, IDATA stage, void* reserved) {
 			TRIGGER_J9HOOK_VM_ABOUT_TO_BOOTSTRAP(vm->hookInterface, vm->mainThread);
 			/* At this point, the decision about which interpreter to use has been made */
 			vm->bytecodeLoop = J9_ARE_ANY_BITS_SET(vm->extendedRuntimeFlags, J9_EXTENDED_RUNTIME_DEBUG_MODE)
-					? (void*)debugBytecodeLoop
-					: (void*)bytecodeLoop;
+					? debugBytecodeLoop
```
Comment (gacholio) [inline runtime/vm/jvminit.c:2435]: This will likely fail to compile on other platforms. Not clear why this change is necessary.

Comment (ChengJin01) [reply runtime/vm/jvminit.c:2435]: @gacholio ,  the change here is the key to call the interpreter in /runtime/vm/rv64cinterpnojit.cpp
```
void
cInterpreter(J9VMThread *currentThread)
{
	J9JavaVM *vm = currentThread->javaVM;
	vm->bytecodeLoop(currentThread); <------
}
```
Otherwise, the VM will totally fail to start.  Technically it might be correct to directly call bytecodeLoop() here like by modi:
```
void
cInterpreter(J9VMThread *currentThread)
{
     bytecodeLoop(currentThread); <------
}
``
But I am not too sure whether there is any issue introduced with this approach.

Comment (gacholio) [reply runtime/vm/jvminit.c:2435]: I still don't see why you need to remove the cast. And no, you can't call it directly - it has to be called indirect to support debug (and the upcoming mixed mode).

Comment (ChengJin01) [reply runtime/vm/jvminit.c:2435]: I still don't see why you need to remove the cast. <---- this change was proposed previously by @keithc-ca  as follows:
```
According to vm_api.h [1] those functions have signature UDATA (*)(J9VMThread *) while the assignment in jvminit.c expects this field to be of type void * [2]: I don't expect this to compile without further changes.

[1] https://github.com/eclipse/openj9/blob/master/runtime/oti/vm_api.h#L4069-L4080
[2] https://github.com/eclipse/openj9/blob/master/runtime/vm/jvminit.c#L2414-L2416
```

Comment (gacholio) [reply runtime/vm/jvminit.c:2435]: Sorry, didn't see that you had also changed the type in the VM structure. I agree this is more correct, otherwise you'd have to do the cast before the call.

**Comment by ChengJin01** [comment]
@keithc-ca & @gacholio ,  I updated the code with bytecodeLoop accordingly to to avoid conflict as @sharon-wang just merged her changes for split interpreter. Please double-check in case any issue is introduced.

**Comment by gacholio** [comment]
The jvminit merge seems correct to me.

**Comment by sharon-wang** [comment]
Agreed, bytecodeLoop updates look good!

File: runtime/oti/j9nonbuilder.h [hunk 6]
```diff
@@ -2511,7 +2544,40 @@ typedef struct J9SFJ2IFrame {
...
+	UDATA* jit_r22;
+	UDATA* jit_r23;
+	UDATA* jit_r24;
+	UDATA* jit_r25;
+	UDATA* jit_r26;
+	UDATA* jit_r27;
+	UDATA* jit_r28;
+	UDATA* jit_r29;
+	UDATA* jit_r30;
+	UDATA* jit_r31;
```
Comment (keithc-ca) [inline runtime/oti/j9nonbuilder.h:2579]: Why are these fields pointers (`UDATA *`) unlike other platforms which use `UDATA`?

Comment (ChengJin01) [reply runtime/oti/j9nonbuilder.h]: Fixed to align with the definition of JIT registers on other platforms.

**Comment by keithc-ca** [comment]
@gacholio Please add your approval if you're satisfied with this.

File: runtime/vm/rv64cinterpnojit.cpp [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+#include "BytecodeInterpreter.hpp"
+#include "ut_j9vm.h"
+
+/* This code here should be removed once the JIT related code
+ * is implemented in rv64cinterp.m4.
+ */
+extern "C" {
+
+void
```
Comment (gacholio) [inline runtime/vm/rv64cinterpnojit.cpp:31]: This function should be removed - since you have the wrapper m4 file, you may as well declare the symbol there.

Comment (ChengJin01) [reply runtime/vm/rv64cinterpnojit.cpp:31]: @gacholio , the mp4 file (not working for now) is an empty file as a placeholder to be modified by JIT to add JIT related assembly in there.

File: runtime/vm/rv64cinterpnojit.cpp [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+cInterpreter(J9VMThread *currentThread)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	vm->bytecodeLoop(currentThread);
+}
+
+void
```
Comment (gacholio) [inline runtime/vm/rv64cinterpnojit.cpp:38]: This function is incorrect - it should be calling vm->bytecodeLoop and switching on the result, not switching before call.

Comment (gacholio) [reply runtime/vm/rv64cinterpnojit.cpp:38]: If the result is J9_BCLOOP_EXIT_INTERPRETER then return, otherwise assert.

File: runtime/vm/rv64cinterpnojit.cpp (hunk 7 above)
Comment (ChengJin01) [reply runtime/vm/rv64cinterpnojit.cpp:31]: As discussed with @gacholio offline, vm->bytecodeLoop(currentThread) should be directly called via c_cInterpreter while cInterpreter is moved to the .mp4 file to be added along with JIT code there.

File: runtime/vm/rv64cinterpnojit.cpp (hunk 8 above)
Comment (ChengJin01) [reply runtime/vm/rv64cinterpnojit.cpp]: Updated the all code including `rv64cinterp.m4` as suggested above.

**Comment by gacholio** [comment]
jenkins test sanity zlinux jdk8

**Comment by gacholio** [comment]
jenkins test sanity plinux jdk8

**Comment by gacholio** [comment]
I saw enough testing pass to prove there's no cross-platform issues, and the latest push was an obvious fix in a file only compiled on the new processor, so I'm merging this.
